amads/time/meter/__init__.py

```diff
@@ -0,0 +1,619 @@
+"""
+This module serves to map out metrical hierarchies in a number of different ways and
+to express the relationship between
+the notes in
+and
+the hierarchy of
+a metrical cycle.
+
+Uses include identifying notes that traverse metrical levels,
+for analysis (e.g., as a measure of syncopation)
+and notation (e.g., re-notating to reflect the
+within-measure notational conventions).
+"""
+
+from typing import Optional, Union
+
+# ------------------------------------------------------------------------------
+
+
+class MetricalHierarchy:
+    """
+    This class serves to create standard, interoperable representations
+    of metrical hierarchy,
+    centred on a single representation
+    of metrical levels in terms of starts expressed by quarter length
+    from the start of the measure,
+    and (optionally) other representations.
+
+    At least one of the parameters (below) must be specified.
+    These are listed in order of override, e.g., if a full `start_hierarchy` is specified, all else are ignored.
+
+    For a 'quick start', the `time_signature` defaults will serve most use cases, and
+    that along with `levels` should be cover almost all.
+
+    Parameters
+    ----------
+    start_hierarchy
+        Users can specify the `start_hierarchy` directly
+        and completely from scratch (ignoring all other parameters and defaults).
```

Comment:
It would be good to include a more formal definition of start_hierarchy here. I'm not entirely sure about whether it's good to include 4.0 as the final element in 4/4, for example, because in my mind 4.0 is equal to 0.0 in the relevant modular arithmetic

Reply:
An alternative would be to encode bar length (or cycle length??) as a separate attribute of `MetricalHierarchy`

Reply:
Yes, let's have a `self.cycle_length = self.start_hierarchy[0][-1]` on the init. I'll give that its own commit.

Whether to include the full cycle value is open to discussion. Some routines require it but can be handled otherwise as you suggest.